presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java

```diff
@@ -0,0 +1,113 @@
+import java.util.Date;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static io.airlift.http.client.Request.Builder.fromRequest;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class InternalAuthenticationManager
+        implements HttpRequestFilter
+{
+    private static final Logger log = Logger.get(InternalAuthenticationManager.class);
+
+    private static final String PRESTO_INTERNAL_BEARER = "X-Presto-Internal-Bearer";
+
+    private final Optional<JwtParser> jwtParser;
+    private final Optional<Supplier<String>> jwtGenerator;
+
+    @Inject
+    public InternalAuthenticationManager(InternalCommunicationConfig internalCommunicationConfig, NodeInfo nodeInfo)
+    {
+        Optional<String> sharedSecret = requireNonNull(internalCommunicationConfig, "internalCommunicationConfig is null").getSharedSecret();
+        if (sharedSecret.isPresent()) {
+            byte[] hmac = Hashing.sha256().hashString(sharedSecret.get(), UTF_8).asBytes();
+            this.jwtParser = Optional.of(Jwts.parser().setSigningKey(hmac));
+            String nodeId = nodeInfo.getNodeId();
+            this.jwtGenerator = Optional.of(() -> generateJwt(hmac, nodeId));
+        }
+        else {
+            this.jwtParser = Optional.empty();
+            this.jwtGenerator = Optional.empty();
+        }
+    }
+
+    private static String generateJwt(byte[] hmac, String nodeId)
+    {
+        return Jwts.builder()
+                .signWith(SignatureAlgorithm.HS256, hmac)
+                .setSubject(nodeId)
+                .setExpiration(Date.from(ZonedDateTime.now().plusMinutes(5).toInstant()))
+                .compact();
+    }
+
+    public boolean isInternalRequest(HttpServletRequest request)
+    {
+        return request.getHeader(PRESTO_INTERNAL_BEARER) != null;
+    }
+
+    public Principal authenticateInternalRequest(HttpServletRequest request)
+    {
+        if (!jwtParser.isPresent()) {
+            log.error("Internal authentication in not configured");
+            return null;
+        }
+
+        String internalBarer = request.getHeader(PRESTO_INTERNAL_BEARER);
+        try {
+            Jws<Claims> claimsJws = jwtParser.get().parseClaimsJws(internalBarer);
+            String subject = claimsJws.getBody().getSubject();
+            return new InternalPrincipal(subject);
+        }
+        catch (JwtException e) {
+            log.error(e, "Internal authentication failed");
+            return null;
+        }
+        catch (RuntimeException e) {
+            throw new RuntimeException("Authentication error", e);
+        }
+    }
+
+    @Override
+    public Request filterRequest(Request request)
+    {
+        return jwtGenerator.map(Supplier::get)
+                .map(jwt -> fromRequest(request)
+                        .addHeader(PRESTO_INTERNAL_BEARER, jwt)
+                        .build())
+                .orElse(request);
```

Comment:
wouldn't this be more readable?
```
if (!jwtGenerator.isPresent()) {
    return request;
}

return fromRequest(request)
    .addHeader(PRESTO_INTERNAL_BEARER, jwtGenerator.get().get())
    .build();
```

Reply:
This generates an intellij warning. That is the simplest code found for this